### PR TITLE
Convert the property value from the platform encoding to MUTF8 string

### PR DESCRIPTION
**Pull request by ChengJin01**
The change is to convert the property value string from the platform
encoding to the modified UTF8 to adapt to the OS locale setting.

Fixes: #11139

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Reviewer: @pshipton 
FYI: @gacholio, @tajila

File: runtime/vm/vmprops.c [hunk 1]
```diff
@@ -1040,13 +1040,22 @@ getJ9VMVersionString(J9JavaVM *vm) {
...
 	J9VMSystemProperty* newProp = pool_newElement(vm->systemProperties);
 	if (NULL == newProp) {
 		return J9SYSPROP_ERROR_OUT_OF_MEMORY;
 	}
 
+	/* Convert the value string from the platform encoding to the modified UTF8 */
+	convertedPropValue = (char *)getMUtf8String(vm, propValue, strlen(propValue));
+	if (NULL == convertedPropValue) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		j9mem_free_memory(newProp);
```
Comment (gacholio) [inline runtime/vm/vmprops.c:1053]: This is incorrect - should be `pool_removeElement`.

File: runtime/vm/vmprops.c [hunk 2]
```diff
@@ -1040,13 +1040,22 @@ getJ9VMVersionString(J9JavaVM *vm) {
...
 addSystemProperty(J9JavaVM * vm, const char* propName,  const char* propValue, UDATA flags)
 {
+	char *convertedPropValue = NULL;
 	J9VMSystemProperty* newProp = pool_newElement(vm->systemProperties);
 	if (NULL == newProp) {
 		return J9SYSPROP_ERROR_OUT_OF_MEMORY;
 	}
 
+	/* Convert the value string from the platform encoding to the modified UTF8 */
+	convertedPropValue = (char *)getMUtf8String(vm, propValue, strlen(propValue));
```
Comment (pshipton) [inline runtime/vm/vmprops.c:1050]: This is going to be a little trickier. Sometimes the arguments to addSystemProperty() are already converted to utf8.
See https://github.com/eclipse/openj9/blob/master/runtime/vm/vmprops.c#L856-L861 
and https://github.com/eclipse/openj9/blob/master/runtime/vm/vmprops.c#L907

Either the callers of addSystemProperty() need to pass in a new argument to indicate the value should be converted, or the conversion needs to be done elsewhere before addSystemProperty() is called.

File: runtime/vm/vmprops.c (hunk 1 above)
Comment (ChengJin01) [reply runtime/vm/vmprops.c:1053]: The fix here will to be moved out as suggested by Pete.

File: runtime/vm/vmprops.c (hunk 2 above)
Comment (ChengJin01) [reply runtime/vm/vmprops.c:1050]: As discussed with @pshipton offline, the fix above will be moved to `addPropertiesForOptionWithAssignArg` and all functions in addModularitySystemProperties including `addPropertiesForOptionWithAssignArg` should be fixed in the same way to address all these cases for the platform setting.

Comment (ChengJin01) [reply runtime/vm/vmprops.c]: The code has been updated at https://github.com/eclipse/openj9/commit/30329871ee1a60edf626ffa8e542638851913045. Will compile a build locally with the fix to see whether it still works good as expected.

File: runtime/vm/vmprops.c [hunk 3]
```diff
@@ -152,6 +152,11 @@ getOptionArg(J9JavaVM *vm, IDATA argIndex, UDATA optionNameLen)
 	}
 
 _end:
+	if (NULL != optionArg) {
+		/* Convert the value string from the platform encoding to the modified UTF8 */
+		optionArg = (char *)getMUtf8String(vm, optionArg, strlen(optionArg));
```
Comment (pshipton) [inline runtime/vm/vmprops.c:157]: The places that use this need to pass the J9SYSPROP_FLAG_VALUE_ALLOCATED flag so the memory is freed.

Comment (pshipton) [reply runtime/vm/vmprops.c:157]: and addPropertyForOptionWithModuleListArg() should free the memory.

Comment (pshipton) [reply runtime/vm/vmprops.c:157]: Also, pls add a function comment for this function which mentions the return value is allocated and needs to be freed.

Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: Updated the code against the suggestions above.

Comment (pshipton) [reply runtime/vm/vmprops.c:159]: addPropertyForOptionWithModuleListArg() needs to free the memory, and addPropertiesForOptionWithAssignArg needs to pass J9SYSPROP_FLAG_VALUE_ALLOCATED as well.

Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: For `addPropertiesForOptionWithAssignArg`, the flag is already in there (not in our fix)
```
addPropertiesForOptionWithAssignArg(J9JavaVM *vm, const char *optionName, UDATA optionNameLen, const char *basePropName, UDATA basePropNameLen, UDATA *propertyCount)
{
				rc = addSystemProperty(vm, propName, optionArg, J9SYSPROP_FLAG_NAME_ALLOCATED); <----
				if (J9SYSPROP_ERROR_NONE != rc) {
					goto _end;
				}

```

Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: Added code `j9mem_free_memory(optionArg)` in the loop in `addPropertyForOptionWithModuleListArg()`.

Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: add J9SYSPROP_FLAG_VALUE_ALLOCATED to `addPropertiesForOptionWithAssignArg` given only `J9SYSPROP_FLAG_NAME_ALLOCATED` is passed to `addSystemProperty` in there.

File: runtime/vm/vmprops.c [hunk 4]
```diff
@@ -309,6 +316,8 @@ addPropertyForOptionWithModuleListArg(J9JavaVM *vm, const char *optionName, IDAT
 					}
 				}
 				CONSUME_ARG(j9vm_args, argIndex);
+
+				j9mem_free_memory(optionArg);
```
Comment (pshipton) [inline runtime/vm/vmprops.c:320]: Above there is a case that does `goto _end` which needs to free optionArg

Comment (pshipton) [reply runtime/vm/vmprops.c:320]: two cases actually

File: runtime/vm/vmprops.c (hunk 3 above)
Comment (pshipton) [reply runtime/vm/vmprops.c:159]: Ok, and pls add a function comment for this function which mentions the return value is allocated and needs to be freed.

Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: Added comment for `optionArg` in `addPropertiesForOptionWithAssignArg`.

File: runtime/vm/vmprops.c (hunk 4 above)
Comment (ChengJin01) [reply runtime/vm/vmprops.c:322]: Fixed the cases as mentioned above.

File: runtime/vm/vmprops.c (hunk 3 above)
Comment (pshipton) [reply runtime/vm/vmprops.c:159]: I'm asking for a function comment to be added to getOptionArg().

File: runtime/vm/vmprops.c [hunk 5]
```diff
@@ -370,6 +381,7 @@ addPropertiesForOptionWithAssignArg(J9JavaVM *vm, const char *optionName, UDATA
 		UDATA indexLen = 1;
 
 		do {
+			/* The allocated return value for optionArg needs to be freed after use */
```
Comment (pshipton) [inline runtime/vm/vmprops.c:384]: There are 4 places calling this, why only comment one of them?

File: runtime/vm/vmprops.c (hunk 3 above)
Comment (ChengJin01) [reply runtime/vm/vmprops.c:159]: Added the function comment to `getOptionArg()` as suggested above.

File: runtime/vm/vmprops.c (hunk 5 above)
Comment (ChengJin01) [reply runtime/vm/vmprops.c]: Already added the function comment for `getOptionArg`.

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdk11

**Comment by pshipton** [comment]
jenkins test sanity.system zlinux jdk11

**Comment by pshipton** [comment]
jenkins test sanity.system win jdk11

**Comment by pshipton** [comment]
jenkins test sanity.system zlinux jdk11

**Comment by ChengJin01** [comment]
Verified with the latest code which works good
```
$ export LANG=zh_CN
$ locale
LANG=zh_CN
LC_COLLATE="zh_CN"
LC_CTYPE="zh_CN"
LC_MONETARY="zh_CN"
LC_NUMERIC="zh_CN"
LC_TIME="zh_CN"
LC_MESSAGES="zh_CN"
LC_ALL=
$ ./test_zh_CN.sh
+ javac -d mod module-info.java test.java
+ javac -p mod --add-modules ??_module test0.java
+ java -p mod --add-modules ??_module test0
method1


$ export LANG=ja_JP
$ locale
LANG=ja_JP
LC_COLLATE="ja_JP"
LC_CTYPE="ja_JP"
LC_MONETARY="ja_JP"
LC_NUMERIC="ja_JP"
LC_TIME="ja_JP"
LC_MESSAGES="ja_JP"
LC_ALL=
$ ./test_ja_JP_euc.sh
+ javac -d mod module-info.java test.java
+ javac -p mod --add-modules ??_module test0.java
+ java -p mod --add-modules ??_module test0
method1



$ export LANG=Ja_JP
$ locale
LANG=Ja_JP
LC_COLLATE="Ja_JP"
LC_CTYPE="Ja_JP"
LC_MONETARY="Ja_JP"
LC_NUMERIC="Ja_JP"
LC_TIME="Ja_JP"
LC_MESSAGES="Ja_JP"
LC_ALL=
$  ./test_Ja_JP.sh
+ export LANG=Ja_JP
+ javac -d mod module-info.java test.java
+ javac -p mod --add-modules ??_module test0.java
+ java -p mod --add-modules ??_module test0
method1

```

**Review by pshipton — APPROVED** [review]

**Comment by pshipton** [comment]
@gacholio any further concerns?

**Review by gacholio — APPROVED** [review]

**Pull request merged**